File: bindings/assemblyscript/src/safe_math.ts
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

/**
 * SafeMath - Overflow-safe arithmetic operations.
 *
 * Thin wrapper around libproven's SafeMath functions. All arithmetic
 * computation is performed by the formally verified Idris 2 implementation
 * via host-provided WASM imports. This module only handles data marshaling.
 *
 * Corresponds to the SafeMath section in proven.h:
 *   - proven_math_div, proven_math_mod
 *   - proven_math_add_checked, proven_math_sub_checked, proven_math_mul_checked
 *   - proven_math_abs_safe, proven_math_clamp, proven_math_pow_checked
 */

import { Result, RESULT_BUF, readIntResult } from "./common";
import {
  proven_math_div,
  proven_math_mod,
  proven_math_add_checked,
  proven_math_sub_checked,
  proven_math_mul_checked,
  proven_math_abs_safe,
  proven_math_clamp,
  proven_math_pow_checked,
} from "./ffi";

/**
 * SafeMath provides overflow-checked arithmetic operations.
 *
 * Every method delegates to the corresponding libproven host function.
 * No arithmetic logic is implemented in AssemblyScript.
 */
export class SafeMath {
  /**
   * Checked addition with overflow detection.
   * Delegates to proven_math_add_checked (Idris 2 verified).
   *
   * @param a - First operand.
   * @param b - Second operand.
   * @returns Result with sum, or ErrOverflow if overflow would occur.
   */
  static checkedAdd(a: i64, b: i64): Result<i64> {
    proven_math_add_checked(a, b, RESULT_BUF);
    return readIntResult(RESULT_BUF);
  }

  /**
   * Checked subtraction with underflow detection.
   * Delegates to proven_math_sub_checked (Idris 2 verified).
   *
   * @param a - First operand.
   * @param b - Second operand.
   * @returns Result with difference, or ErrUnderflow if underflow would occur.
   */
  static checkedSub(a: i64, b: i64): Result<i64> {
    proven_math_sub_checked(a, b, RESULT_BUF);
    return readIntResult(RESULT_BUF);
  }

  /**
   * Checked multiplication with overflow detection.
   * Delegates to proven_math_mul_checked (Idris 2 verified).
   *
   * @param a - First operand.
   * @param b - Second operand.
   * @returns Result with product, or ErrOverflow if overflow would occur.
   */
  static checkedMul(a: i64, b: i64): Result<i64> {
    proven_math_mul_checked(a, b, RESULT_BUF);
    return readIntResult(RESULT_BUF);
  }

  /**
   * Safe integer division.
   * Delegates to proven_math_div (Idris 2 verified).
   *
   * @param numerator - Dividend.
   * @param denominator - Divisor.
   * @returns Result with quotient, or ErrDivisionByZero/ErrOverflow.
   */
  static checkedDiv(numerator: i64, denominator: i64): Result<i64> {
    proven_math_div(numerator, denominator, RESULT_BUF);
    return readIntResult(RESULT_BUF);
  }

  /**
   * Safe modulo operation.
   * Delegates to proven_math_mod (Idris 2 verified).
   *
   * @param numerator - Dividend.
   * @param denominator - Divisor.
   * @returns Result with remainder, or ErrDivisionByZero.
   */
  static checkedMod(numerator: i64, denominator: i64): Result<i64> {
    proven_math_mod(numerator, denominator, RESULT_BUF);
    return readIntResult(RESULT_BUF);
  }

  /**
   * Safe absolute value.
   * Delegates to proven_math_abs_safe (Idris 2 verified).
   *
   * @param n - Value to take absolute value of.
   * @returns Result with absolute value, or ErrOverflow for i64.MIN_VALUE.
   */
  static checkedAbs(n: i64): Result<i64> {
    proven_math_abs_safe(n, RESULT_BUF);
    return readIntResult(RESULT_BUF);
  }

  /**
   * Clamp value to [lo, hi] range.
   * Delegates to proven_math_clamp (Idris 2 verified).
   *
   * This is infallible -- always returns a valid i64.
   *
   * @param lo - Lower bound (inclusive).
   * @param hi - Upper bound (inclusive).
   * @param value - Value to clamp.
   * @returns Clamped value.
   */
  static clamp(lo: i64, hi: i64, value: i64): i64 {
    return proven_math_clamp(lo, hi, value);
  }

  /**
   * Integer exponentiation with overflow checking.
   * Delegates to proven_math_pow_checked (Idris 2 verified).
   *
   * @param base - Base value.
   * @param exp - Exponent (non-negative).
   * @returns Result with power, or ErrOverflow if overflow would occur.
   */
  static checkedPow(base: i64, exp: u32): Result<i64> {
    proven_math_pow_checked(base, exp, RESULT_BUF);
    return readIntResult(RESULT_BUF);
  }
}

File: bindings/ephapax-affine/src/safe_math.rs
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

//! Safe arithmetic operations with affine type semantics.
//!
//! All computation is delegated to libproven's formally verified Idris 2
//! core via the FFI bridge. No arithmetic logic is reimplemented here.
//!
//! ## Affine Type Semantics
//!
//! Each result value can be consumed at most once. The Rust ownership
//! system enforces this: once you match on a `ProvenResult<i64>`, the
//! inner value is moved and cannot be reused without explicit cloning.
//! This mirrors Ephapax's affine type system where resources are
//! consumed at most once.

use crate::ffi;
use crate::{int_result_to_proven, ProvenResult};

/// Affine wrapper for a checked integer value.
///
/// `#[affine]` -- this value can be consumed at most once.
/// Once unwrapped, the inner `i64` is moved out.
#[derive(Debug)]
pub struct AffineInt {
    value: i64,
}

impl AffineInt {
    /// Consume this affine value, returning the inner integer.
    ///
    /// After calling this, the `AffineInt` is consumed and cannot be reused.
    pub fn consume(self) -> i64 {
        self.value
    }
}

/// Checked addition with overflow detection.
///
/// Returns the sum as an affine value, or error on overflow.
/// Delegates to `proven_math_add_checked` via FFI.
pub fn safe_add(a: i64, b: i64) -> ProvenResult<AffineInt> {
    // SAFETY: proven_math_add_checked is a pure computation in Idris 2.
    let result = unsafe { ffi::proven_math_add_checked(a, b) };
    int_result_to_proven(result).map(|v| AffineInt { value: v })
}

/// Checked subtraction with underflow detection.
///
/// Returns the difference as an affine value, or error on underflow.
/// Delegates to `proven_math_sub_checked` via FFI.
pub fn safe_sub(a: i64, b: i64) -> ProvenResult<AffineInt> {
    // SAFETY: proven_math_sub_checked is a pure computation in Idris 2.
    let result = unsafe { ffi::proven_math_sub_checked(a, b) };
    int_result_to_proven(result).map(|v| AffineInt { value: v })
}

/// Checked multiplication with overflow detection.
///
/// Returns the product as an affine value, or error on overflow.
/// Delegates to `proven_math_mul_checked` via FFI.
pub fn safe_mul(a: i64, b: i64) -> ProvenResult<AffineInt> {
    // SAFETY: proven_math_mul_checked is a pure computation in Idris 2.
    let result = unsafe { ffi::proven_math_mul_checked(a, b) };
    int_result_to_proven(result).map(|v| AffineInt { value: v })
}

/// Safe division with division-by-zero protection.
///
/// Returns the quotient as an affine value, or error on division by zero.
/// Delegates to `proven_math_div` via FFI.
pub fn safe_div(numerator: i64, denominator: i64) -> ProvenResult<AffineInt> {
    // SAFETY: proven_math_div is a pure computation in Idris 2.
    let result = unsafe { ffi::proven_math_div(numerator, denominator) };
    int_result_to_proven(result).map(|v| AffineInt { value: v })
}

/// Safe modulo with division-by-zero protection.
///
/// Returns the remainder as an affine value, or error on zero divisor.
/// Delegates to `proven_math_mod` via FFI.
pub fn safe_mod(numerator: i64, denominator: i64) -> ProvenResult<AffineInt> {
    // SAFETY: proven_math_mod is a pure computation in Idris 2.
    let result = unsafe { ffi::proven_math_mod(numerator, denominator) };
    int_result_to_proven(result).map(|v| AffineInt { value: v })
}

/// Safe absolute value (handles MIN_INT correctly).
///
/// Returns the absolute value as an affine value, or error for i64::MIN.
/// Delegates to `proven_math_abs_safe` via FFI.
pub fn safe_abs(n: i64) -> ProvenResult<AffineInt> {
    // SAFETY: proven_math_abs_safe is a pure computation in Idris 2.
    let result = unsafe { ffi::proven_math_abs_safe(n) };
    int_result_to_proven(result).map(|v| AffineInt { value: v })
}

/// Safe negation (handles MIN_INT correctly).
///
/// Returns the negated value as an affine value, or error for i64::MIN.
/// Delegates to `proven_math_negate_safe` via FFI.
pub fn safe_negate(n: i64) -> ProvenResult<AffineInt> {
    // SAFETY: proven_math_negate_safe is a pure computation in Idris 2.
    let result = unsafe { ffi::proven_math_negate_safe(n) };
    int_result_to_proven(result).map(|v| AffineInt { value: v })
}

/// Clamp value to range [lo, hi].
///
/// Always succeeds (no error possible). The result is affine.
/// Delegates to `proven_math_clamp` via FFI.
pub fn clamp(lo: i64, hi: i64, value: i64) -> AffineInt {
    // SAFETY: proven_math_clamp always succeeds.
    let v = unsafe { ffi::proven_math_clamp(lo, hi, value) };
    AffineInt { value: v }
}

/// Checked exponentiation with overflow detection.
///
/// Returns the result as an affine value, or error on overflow.
/// Delegates to `proven_math_pow_checked` via FFI.
pub fn safe_pow(base: i64, exp: u32) -> ProvenResult<AffineInt> {
    // SAFETY: proven_math_pow_checked is a pure computation in Idris 2.
    let result = unsafe { ffi::proven_math_pow_checked(base, exp) };
    int_result_to_proven(result).map(|v| AffineInt { value: v })
}
